### ISSUE TITLE
Body is located before Header (Unordered tags)

### ISSUE THREAD

**Issue by Denis-Svichkarev** [gcharita/XMLMapper#18]
Hello, I am trying to compose a SOAP request and have encountered a problem - for some reason in my request Body is located before Header and I don't understand how I can change it. 

This is how I create SOAPEnvelope where is **information** has header data and **soapMessage** has body data:
`let soapEnvelope = SOAPEnvelope(soapMessage: soapMessage, soapInformation: information, soapVersion: .version1point2)`

And for example when I try to print SOAPEnvelope (soapEnvelope.toXMLString()) I receive this string:

`<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope soap:encodingStyle="http://www.w3.org/2003/05/soap-encoding" xmlns:soap="http://www.w3.org/2003/05/soap-envelope">
<soap:Body>
		....
</soap:Body>
<soap:Header>
		...
</soap:Header>
</soap:Envelope>`

The problem is that I work with the ONVIF service and it is sensitive to the order of the header / body.

Thanks

**Comment by Denis-Svichkarev** [comment]
I tried to find the cause and realized that this was due to the formation of the XML string using a dictionary in which the objects are out of order. Then I got to **Dictionary+XMLParser.swift** and maybe, as a temporary solution, made a manual correction of the header/body order:

`var innerXML: String {
        var nodes: [String] = []
        
        comments?.forEach({ (comment: String) in
            nodes.append(String(format: "", comment.xmlEncodedString))
        })
        
        childNodes?.forEach({ (childNode:(key: String, value: Any)) in
            if let xmlStringNode = XMLParserHelper.xmlString(forNode: childNode.value, withNodeName: childNode.key) {
                nodes.append(xmlStringNode)
            }
        })
        
        if let text = innerText {
            nodes.append(text)
        }
        
        // ---- temporary fix ----
        var orderedRootNodes: [String] = []
        if nodes.count == 2 {
            if nodes[0].contains("Body") && nodes[1].contains("Header") {
                orderedRootNodes.append(nodes[1])
                orderedRootNodes.append(nodes[0])
                return orderedRootNodes.joined(separator: "\n")
            }
        }
        // ---------------------
        
        return nodes.joined(separator: "\n")
    }
`

**Comment by gcharita** [comment]
This is an issue. Tag ordering are important in many cases.

Since Swift version 4.2, [Dictionary](https://developer.apple.com/documentation/swift/dictionary) are actually unordered 

> Every dictionary is an unordered collection of key-value pairs.

Did your temporary solution worked?

**Comment by Denis-Svichkarev** [comment]
Yes, it works

**Comment by gcharita** [comment]
This is good, but we need to think of a generic solution to this problem. Please, keep this issue open.
Thanks for reporting it.

**Comment by gcharita** [comment]
A fix for this issue will be included in the next release (v1.5.2)

You can now use `nodesOrder` property of `XMLMap` to preserve or change the nodes ordering:

```swift
class TestOrderedNodes: XMLMappable {
    var nodeName: String!

    var id: String?
    var name: String?
    var nodesOrder: [String]?

    init() {}
    required init?(map: XMLMap) {}

    func mapping(map: XMLMap) {
        id <- map["id"]
        name <- map["name"]
        nodesOrder <- map.nodesOrder
    }
}

let testOrderedNodes = TestOrderedNodes()
testOrderedNodes.id = "1"
testOrderedNodes.name = "the name"
testOrderedNodes.nodesOrder = ["id", "name"]
print(testOrderedNodes.toXMLString() ?? "nil")
```

Of course `SOAPEnvelope` will have by default the correct ordering (first the soap:Header and then the soap:Body)
